storage/src/index/partitioned/unordered.rs

```diff
@@ -0,0 +1,276 @@
+//! The unordered variant of a partitioned index.
+
+use crate::{index::Unordered, translator::Translator};
+use commonware_runtime::Metrics;
+use std::marker::PhantomData;
+
+/// A partitioned index that maps translated keys to values. The first `P` bytes of the
+/// (untranslated) key are used to determine the partition, and the translator is used by the
+/// partition-specific indices on the key after stripping this prefix. The value of `P` should be
+/// small, typically 1 or 2. Anything larger than 3 will fail to compile.
+pub struct Index<T: Translator, I: Unordered<T>, const P: usize> {
+    partitions: Vec<I>,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: Translator, I: Unordered<T>, const P: usize> Index<T, I, P> {
+    /// Create a new [Index] with the given translator.
+    pub fn new(ctx: impl Metrics, translator: T) -> Self {
+        let partition_count = 1 << (P * 8);
+        let mut partitions = Vec::with_capacity(partition_count);
+        for i in 0..partition_count {
+            partitions.push(I::init(
+                ctx.with_label(&format!("partition_{i}")),
+                translator.clone(),
+            ));
+        }
+
+        Self {
+            partitions,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Get the partition index for the given key, along with the prefix-stripped key for probing
+    /// the referenced partition. The returned index value is in the range `[0, 2^(P*8) - 1]`.
+    fn partition_index_and_sub_key(key: &[u8]) -> (usize, &[u8]) {
+        // TODO: Re-evaluate assertion placement after `generic_const_exprs` is stable.
+        const {
+            assert!(P > 0, "P must be greater than 0");
+            assert!(P <= 3, "P must be 3 or less");
+        }
+        let copy_len = P.min(key.len());
+
+        // Because the prefix length has a max of 3, we can safely use a 4-byte array for converting
+        // it into a u32 index.
+        let mut bytes = [0u8; 4];
```

Comment:
nit: wonder if it makes sense to use a `const` for 4 here

Reply:
made it a const.